File: src/utils/tracker.py
"""SQLite-based import tracking for DAM."""

import sqlite3
from pathlib import Path
from datetime import datetime
from typing import Optional
from dataclasses import dataclass


@dataclass
class ImportedAsset:
    """Represents an imported asset record."""

    icloud_uuid: str
    immich_id: Optional[str]
    filename: str
    file_size: int
    media_type: str  # 'photo' or 'video'
    imported_at: datetime


class ImportTracker:
    """Track which iCloud assets have been imported to Immich."""

    def __init__(self, db_path: Path):
        self.db_path = db_path
        self._conn: Optional[sqlite3.Connection] = None
        self._ensure_schema()

    def _get_conn(self) -> sqlite3.Connection:
        """Get or create database connection."""
        if self._conn is None:
            self._conn = sqlite3.connect(self.db_path)
            self._conn.row_factory = sqlite3.Row
        return self._conn

    def _ensure_schema(self) -> None:
        """Create database schema if it doesn't exist."""
        conn = self._get_conn()
        conn.executescript(
            """
            CREATE TABLE IF NOT EXISTS imported_assets (
                icloud_uuid TEXT PRIMARY KEY,
                immich_id TEXT,
                filename TEXT NOT NULL,
                file_size INTEGER NOT NULL,
                media_type TEXT NOT NULL,
                imported_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            );

            CREATE INDEX IF NOT EXISTS idx_imported_at 
                ON imported_assets(imported_at);
            
            CREATE INDEX IF NOT EXISTS idx_media_type 
                ON imported_assets(media_type);
        """
        )
        conn.commit()

    def is_imported(self, icloud_uuid: str) -> bool:
        """Check if an asset has already been imported."""
        conn = self._get_conn()
        cursor = conn.execute(
            "SELECT 1 FROM imported_assets WHERE icloud_uuid = ?", (icloud_uuid,)
        )
        return cursor.fetchone() is not None

    def get_imported_uuids(self) -> set[str]:
        """Get all imported iCloud UUIDs as a set for fast lookup."""
        conn = self._get_conn()
        cursor = conn.execute("SELECT icloud_uuid FROM imported_assets")
        return {row["icloud_uuid"] for row in cursor.fetchall()}

    def mark_imported(
        self,
        icloud_uuid: str,
        immich_id: Optional[str],
        filename: str,
        file_size: int,
        media_type: str,
    ) -> None:
        """Mark an asset as imported."""
        conn = self._get_conn()
        conn.execute(
            """
            INSERT OR REPLACE INTO imported_assets 
                (icloud_uuid, immich_id, filename, file_size, media_type, imported_at)
            VALUES (?, ?, ?, ?, ?, ?)
        """,
            (icloud_uuid, immich_id, filename, file_size, media_type, datetime.now()),
        )
        conn.commit()

    def get_stats(self) -> dict:
        """Get import statistics."""
        conn = self._get_conn()

        # Total count
        total = conn.execute("SELECT COUNT(*) FROM imported_assets").fetchone()[0]

        # By media type
        photos = conn.execute(
            "SELECT COUNT(*) FROM imported_assets WHERE media_type = 'photo'"
        ).fetchone()[0]
        videos = conn.execute(
            "SELECT COUNT(*) FROM imported_assets WHERE media_type = 'video'"
        ).fetchone()[0]

        # Total size
        total_size = (
            conn.execute("SELECT SUM(file_size) FROM imported_assets").fetchone()[0]
            or 0
        )

        # Last import
        last_import = conn.execute(
            "SELECT MAX(imported_at) FROM imported_assets"
        ).fetchone()[0]

        return {
            "total": total,
            "photos": photos,
            "videos": videos,
            "total_size_bytes": total_size,
            "total_size_gb": round(total_size / (1024**3), 2),
            "last_import": last_import,
        }

    def get_recent(self, limit: int = 10) -> list[ImportedAsset]:
        """Get recently imported assets."""
        conn = self._get_conn()
        cursor = conn.execute(
            """
            SELECT * FROM imported_assets 
            ORDER BY imported_at DESC 
            LIMIT ?
        """,
            (limit,),
        )
        return [
            ImportedAsset(
                icloud_uuid=row["icloud_uuid"],
                immich_id=row["immich_id"],
                filename=row["filename"],
                file_size=row["file_size"],
                media_type=row["media_type"],
                imported_at=datetime.fromisoformat(row["imported_at"]),
            )
            for row in cursor.fetchall()
        ]

    def close(self) -> None:
        """Close database connection."""
        if self._conn:
            self._conn.close()
            self._conn = None

    def __enter__(self) -> "ImportTracker":
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.close()

File: swift/Sources/PhotosSync/Database/Tracker.swift
import Foundation
import SQLite3

private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

/// Tracks imported assets - shares SQLite DB with Python code
/// Thread-safe via serial dispatch queue
class Tracker {
    private var db: OpaquePointer?
    private let dbPath: URL
    private let queue = DispatchQueue(label: "tracker.db.queue")
    
    init(dbPath: URL) throws {
        self.dbPath = dbPath
        
        if sqlite3_open(dbPath.path, &db) != SQLITE_OK {
            throw TrackerError.openFailed(String(cString: sqlite3_errmsg(db)))
        }
        
        // Enable WAL mode for better concurrent read performance
        var errMsg: UnsafeMutablePointer<CChar>?
        sqlite3_exec(db, "PRAGMA journal_mode=WAL", nil, nil, &errMsg)
        
        try createTables()
    }
    
    deinit {
        sqlite3_close(db)
    }
    
    private func createTables() throws {
        let sql = """
            CREATE TABLE IF NOT EXISTS imported_assets (
                icloud_uuid TEXT PRIMARY KEY,
                immich_id TEXT,
                filename TEXT NOT NULL,
                file_size INTEGER NOT NULL,
                media_type TEXT NOT NULL,
                imported_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                archived INTEGER DEFAULT 0
            );
            
            CREATE INDEX IF NOT EXISTS idx_imported_at ON imported_assets(imported_at);
            CREATE INDEX IF NOT EXISTS idx_media_type ON imported_assets(media_type);
            CREATE INDEX IF NOT EXISTS idx_immich_id ON imported_assets(immich_id);
            CREATE INDEX IF NOT EXISTS idx_archived ON imported_assets(archived);
        """
        
        var errMsg: UnsafeMutablePointer<CChar>?
        if sqlite3_exec(db, sql, nil, nil, &errMsg) != SQLITE_OK {
            let error = errMsg.map { String(cString: $0) } ?? "Unknown error"
            sqlite3_free(errMsg)
            throw TrackerError.execFailed(error)
        }
    }
    
    func isImported(uuid: String) -> Bool {
        let sql = "SELECT 1 FROM imported_assets WHERE icloud_uuid = ? LIMIT 1"
        var stmt: OpaquePointer?
        
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            return false
        }
        defer { sqlite3_finalize(stmt) }
        
        sqlite3_bind_text(stmt, 1, uuid, -1, nil)
        return sqlite3_step(stmt) == SQLITE_ROW
    }
    
    func getImportedUUIDs() -> Set<String> {
        var uuids = Set<String>()
        let sql = "SELECT icloud_uuid FROM imported_assets"
        var stmt: OpaquePointer?
        
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            return uuids
        }
        defer { sqlite3_finalize(stmt) }
        
        while sqlite3_step(stmt) == SQLITE_ROW {
            if let cString = sqlite3_column_text(stmt, 0) {
                uuids.insert(String(cString: cString))
            }
        }
        
        return uuids
    }
    
    func markImported(
        uuid: String,
        immichID: String?,
        filename: String,
        fileSize: Int64,
        mediaType: String
    ) throws {
        try queue.sync {
            let sql = """
                INSERT OR REPLACE INTO imported_assets 
                (icloud_uuid, immich_id, filename, file_size, media_type, imported_at)
                VALUES (?, ?, ?, ?, ?, datetime('now'))
            """
            var stmt: OpaquePointer?
            
            guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
                throw TrackerError.prepareFailed(String(cString: sqlite3_errmsg(db)))
            }
            defer { sqlite3_finalize(stmt) }
            
            sqlite3_bind_text(stmt, 1, uuid, -1, SQLITE_TRANSIENT)
            if let immichID = immichID {
                sqlite3_bind_text(stmt, 2, immichID, -1, SQLITE_TRANSIENT)
            } else {
                sqlite3_bind_null(stmt, 2)
            }
            sqlite3_bind_text(stmt, 3, filename, -1, SQLITE_TRANSIENT)
            sqlite3_bind_int64(stmt, 4, fileSize)
            sqlite3_bind_text(stmt, 5, mediaType, -1, SQLITE_TRANSIENT)
            
            if sqlite3_step(stmt) != SQLITE_DONE {
                throw TrackerError.execFailed(String(cString: sqlite3_errmsg(db)))
            }
        }
    }
    
    func getStats() -> (total: Int, photos: Int, videos: Int, totalBytes: Int64) {
        var total = 0
        var photos = 0
        var videos = 0
        var totalBytes: Int64 = 0
        
        // Total count
        if let count = querySingleInt("SELECT COUNT(*) FROM imported_assets") {
            total = count
        }
        
        // Photos count
        if let count = querySingleInt("SELECT COUNT(*) FROM imported_assets WHERE media_type = 'photo'") {
            photos = count
        }
        
        // Videos count
        if let count = querySingleInt("SELECT COUNT(*) FROM imported_assets WHERE media_type = 'video'") {
            videos = count
        }
        
        // Total size
        if let size = querySingleInt64("SELECT SUM(file_size) FROM imported_assets") {
            totalBytes = size
        }
        
        return (total, photos, videos, totalBytes)
    }
    
    func getImmichIDForUUID(_ uuid: String) -> String? {
        let sql = "SELECT immich_id FROM imported_assets WHERE icloud_uuid = ?"
        var stmt: OpaquePointer?
        
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            return nil
        }
        defer { sqlite3_finalize(stmt) }
        
        sqlite3_bind_text(stmt, 1, uuid, -1, nil)
        
        if sqlite3_step(stmt) == SQLITE_ROW, let cString = sqlite3_column_text(stmt, 0) {
            return String(cString: cString)
        }
        return nil
    }
    
    private func querySingleInt(_ sql: String) -> Int? {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return nil }
        defer { sqlite3_finalize(stmt) }
        
        if sqlite3_step(stmt) == SQLITE_ROW {
            return Int(sqlite3_column_int(stmt, 0))
        }
        return nil
    }
    
    private func querySingleInt64(_ sql: String) -> Int64? {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return nil }
        defer { sqlite3_finalize(stmt) }
        
        if sqlite3_step(stmt) == SQLITE_ROW {
            return sqlite3_column_int64(stmt, 0)
        }
        return nil
    }
    
    /// Mark an asset as archived (deleted from Photos, kept in Immich)
    func markArchived(uuid: String) throws {
        try queue.sync {
            let sql = "UPDATE imported_assets SET archived = 1 WHERE icloud_uuid = ?"
            var stmt: OpaquePointer?
            
            guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
                throw TrackerError.prepareFailed(String(cString: sqlite3_errmsg(db)))
            }
            defer { sqlite3_finalize(stmt) }
            
            sqlite3_bind_text(stmt, 1, uuid, -1, SQLITE_TRANSIENT)
            
            if sqlite3_step(stmt) != SQLITE_DONE {
                throw TrackerError.execFailed(String(cString: sqlite3_errmsg(db)))
            }
        }
    }
    
    /// Check if an asset is archived
    func isArchived(uuid: String) -> Bool {
        let sql = "SELECT archived FROM imported_assets WHERE icloud_uuid = ? LIMIT 1"
        var stmt: OpaquePointer?
        
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            return false
        }
        defer { sqlite3_finalize(stmt) }
        
        sqlite3_bind_text(stmt, 1, uuid, -1, nil)
        
        if sqlite3_step(stmt) == SQLITE_ROW {
            return sqlite3_column_int(stmt, 0) == 1
        }
        return false
    }
    
    /// Get all non-archived imported UUIDs (for import comparison)
    func getActiveImportedUUIDs() -> Set<String> {
        var uuids = Set<String>()
        let sql = "SELECT icloud_uuid FROM imported_assets WHERE archived = 0 OR archived IS NULL"
        var stmt: OpaquePointer?
        
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
            return uuids
        }
        defer { sqlite3_finalize(stmt) }
        
        while sqlite3_step(stmt) == SQLITE_ROW {
            if let cString = sqlite3_column_text(stmt, 0) {
                uuids.insert(String(cString: cString))
            }
        }
        
        return uuids
    }
    
    /// Remove an asset from tracking (when deleted from both Photos and Immich)
    func removeAsset(uuid: String) throws {
        try queue.sync {
            let sql = "DELETE FROM imported_assets WHERE icloud_uuid = ?"
            var stmt: OpaquePointer?
            
            guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else {
                throw TrackerError.prepareFailed(String(cString: sqlite3_errmsg(db)))
            }
            defer { sqlite3_finalize(stmt) }
            
            sqlite3_bind_text(stmt, 1, uuid, -1, SQLITE_TRANSIENT)
            
            if sqlite3_step(stmt) != SQLITE_DONE {
                throw TrackerError.execFailed(String(cString: sqlite3_errmsg(db)))
            }
        }
    }
}

enum TrackerError: Error {
    case openFailed(String)
    case prepareFailed(String)
    case execFailed(String)
}
